search-app/src/cmr/search/services/generic_association_service.clj

```diff
@@ -0,0 +1,259 @@
+  associated_item: {concept_id: C6-PROV1}}]."
+  [context concept associations operation]
+  (let [{:keys [concept-id revision-id]} concept
+        mdb-context (assoc context :system
+                           (get-in context [:system :embedded-systems :metadata-db]))
+        [t1 result] (util/time-execution (util/fast-map
+                                          (fn [association]
+                                            (update-generic-association
+                                             mdb-context
+                                             (merge association
+                                                    {:source-concept-id concept-id
+                                                     :source-revision-id revision-id})
+                                             operation))
+                                          associations))]
+    (info "update-generic-associations:" t1)
+    result))
+
+(defn- fetch-concept
+  "Fetches the the version with revision-id or the latest version of a concept by concept-id with
+  proper error handling."
+  [context concept-type concept-id revision-id]
+  (let [concept (if revision-id
+                  (mdb/get-concept context concept-id revision-id)
+                  (mdb/get-latest-concept context concept-id))]
+    (if-not (empty? (util/remove-nil-keys concept))
+      (if (:deleted concept)
+        (errors/throw-service-error
+         :not-found
+         (format "%s with concept id [%s] was deleted."
+                 (string/capitalize (name concept-type)) concept-id))
+        concept)
+      (errors/throw-service-error
+       :not-found
+       (format "%s could not be found with concept id [%s]"
+               (string/capitalize (name concept-type)) concept-id)))))
+
+(defn- link-to-concepts
+  "Associate/Dissociate a concept to a list of concepts in the associations json
+   based on the given operation type. The operation type can be either :insert or :delete.
+   Throws service error if the concept with the given concept-id and revision-id is not found."
+  [context concept-type concept-id revision-id associations-json operation-type]
+  (let [concept (fetch-concept context concept-type concept-id revision-id)
+        ;; When revision-id is nil, fetch-concept would get the latest revision,
+        ;; in which case (:revision-id concept) = latest revision.
+        ;; Need to set the original revision-id back so that it can be inserted into
+        ;; db correctly.
+        concept (assoc concept :revision-id revision-id)
+        associations (assoc-validation/associations-json->associations associations-json)
+        associations (assoc-validation/validate-no-same-concept-generic-association
+                      concept
+                      associations)
+        associations (assoc-validation/validate-generic-association-types
+                      concept
+                      associations)
```

Comment:
I believe you can combine lines 235-241 with the 'last arg' threading operator ->>, like this:

(->> associations-json
(assoc-validation/associations-json->associations)
(assoc-validation/validate-no-same-concept-generic-association concept)
(assoc-validation/validate-generic-association-types concept))